### ISSUE TITLE
? in battery icon when 100%

### ISSUE THREAD

**Issue by brinzlee** [maxwroc/battery-state-card#111]
**Describe the bug**
I have many battery status entities set up on one page. But most of the time the battery state icon for those that are at 100% will be green with a ? mark inside it. If I refresh the page the icon goes solid green and the ? disappears for a while then returns. I have deleted the Battery State Card a few times in case is was installed incorrectly but I can't work out why it does this. All the other icons that are less than 100% in battery level always show up correctly

**How to reproduce**
As above

**Expected behavior**

**YAML configuration**
```yaml
`color_gradient:
  - '#ff0000'
  - '#ffff00'
  - '#00ff00'
entities:
  - entity: sensor.aeotec_dsb45_water_sensor_battery_level
    name: Garage Aeotec DSB45 Flood Sensor
  - entity: sensor.fibaro_system_fgfs101_zwave_flood_sensor_battery_level
    name: Kitchen FIBARO FGFS101 Flood Sensor
  - entity: sensor.fibaro_system_fgk101_door_opening_sensor_battery_level
    name: Loft Bedroom Window Sensor
  - entity: sensor.fibaro_system_fgk10x_door_opening_sensor_battery_level
    name: Loft Bathroom Window Sensor
  - entity: sensor.fibaro_system_fgms001_zw5_motion_sensor_battery_level
    name: Kitchen Motion Sensor
  - entity: sensor.lumi_remote_b1acn01_32_battery_level
    name: Lounge Smart Switch
  - entity: sensor.mi_magic_cube_battery_level
    name: Magic Cube Sensor
  - entity: sensor.motion_sensor_battery_level
    name: Motion Sensor
  - entity: sensor.multi_sensor_conservatory_battery_level
    name: Conservatory Multi Sensor
  - entity: sensor.multi_sensor_lounge_battery_level
    name: Lounge Multi Sensor
  - entity: sensor.multi_sensor_master_bedroom_battery_level
    name: Master Bedroom Multi Sensor
  - entity: sensor.multi_sensor_olivias_bedroom_battery_level
    name: Olivias Bedroom Multi Sensor
  - entity: sensor.smart_switch_battery_level_2
    name: Master Bedroom Switch
  - entity: sensor.smart_switch_2_battery_level
    name: Olivias Bedroom Switch
  - entity: sensor.vibration_23_battery_level
    name: Vibration Sensor A
  - entity: sensor.vibration_sensor_battery_level
    name: Vibration Sensor B
  - entity: sensor.water_sensor_main_bathroom_battery_level
    name: Main Bathroom Flood
  - entity: sensor.water_sensor_master_bath_suite_battery_level
    name: Master Bathroom Flood
title: Sensor Battery Level Status
type: 'custom:battery-state-card'
`


```

**Screenshots**

**Version**
< 1.4.1>

**Comment by maxwroc** [comment]
People started reporting such issue recently on community forum. Which version of HA do you have? Currently I'm on 0.113.1 and don't see such issues.

Has it started malfunctioning recently? (if you used this card for a while)

I'm installing now HA 0.113.3 because I have started suspecting that this issue might be related to the update somehow

**Comment by brinzlee** [comment]
I've had the issue from the onset of installing it, quite a while ago...and had the problem through all the versions of HA....I'm currently on 0.113.3 too

**Comment by darkkatarsis** [comment]
Im on 115.6, and see this issue for a long time. Even in previous versions.
<img width="1251" alt="Screenshot 2020-10-06 at 08 52 20" src="https://user-images.githubusercontent.com/24244919/95168231-4b5e2900-07b1-11eb-917d-92701c8a2a91.png">
And, after browser refresh:
<img width="1243" alt="Screenshot 2020-10-06 at 08 52 24" src="https://user-images.githubusercontent.com/24244919/95168262-5749eb00-07b1-11eb-97bf-5b21d4f278c3.png">
If I refresh again, it appears "?". Strange...

**Comment by darkkatarsis** [comment]
It can be related to browser cache.

**Comment by brinzlee** [comment]
I don't thinks it's a browser cache issue in my case....Its always the batteries that are marked at 100% that have the question mark

**Comment by maxwroc** [comment]
I agree, this doesn't look like the cache issue. I think that there is some race condition issue or HA calls functions/setters of the card in some weird way (in some cases).

I'm not able to repro this problem still so if one of you guys help me with testing I can try to solve it. I have couple ideas what can be possible problem so I need to check which one is the root cause. Additionally I will add some logging to find out what actually is going on.

Any volunteer?

**Comment by darkkatarsis** [comment]
@maxwroc I can help with no problem

**Comment by maxwroc** [comment]
I'll prepare "debugging" bundle with first attempt of the fix soon (hopefully today in about 2-3h). And I can actually post it here so anyone who is willing to help can give it a try. Stay tuned

**Comment by maxwroc** [comment]
Ok here is the debug version of the card (it is zipped because JS files are not allowed to upload here so please unpack first)
[battery-state-card.zip](https://github.com/maxwroc/battery-state-card/files/5336219/battery-state-card.zip)

Put it wherever you like in the /config/www dir and remember to reference it in the ui-lovelace.yaml. The original card code from HACS must not be present on the page so uninstall it from the HACS (or comment it out in the ui-lovelace.yaml)

Please do not put the battery card on initial lovelace page as you may not be able to repro the issue after refreshing in below steps. Add battery card on one of your pages - temporarily. That page must *not contain more then one battery card* as it will be harder to browse logs.

After changes in ui-lovelace you will need to restart HA (I think it is still required). Open DevTools in your browser - F12 on Win browsers and switch to console tab. Make sure you see the debug version of the card:
![image](https://user-images.githubusercontent.com/8268674/95250546-2a79f000-0812-11eb-9582-01bfb56fdf65.png)

Once everything is setup try to repro the issue (remember to keep dev console open). When you get the repro please click on one of the entries in dev console, Ctrl+A, Ctrl+C, open notepad Ctrl+V, save it and paste this file here.

I didn't have time to add first fix attempt so we have just simple logging here.

**Comment by darkkatarsis** [comment]
<img width="930" alt="Screenshot 2020-10-06 at 23 11 36" src="https://user-images.githubusercontent.com/24244919/95261028-f6aac480-0829-11eb-9565-67456f9c731a.png">
<img width="1789" alt="Screenshot 2020-10-06 at 23 16 14" src="https://user-images.githubusercontent.com/24244919/95261033-f9a5b500-0829-11eb-8791-bafbf73adc9b.png">

log.txt
[log.txt](https://github.com/maxwroc/battery-state-card/files/5336812/log.txt)

**Comment by maxwroc** [comment]
Thanks, and did you go to this page only once? or twice? 

According to the log all the cards were initialized twice.

Ok let's try one more time with first attempt to fix
[battery-state-card.zip](https://github.com/maxwroc/battery-state-card/files/5336997/battery-state-card.zip)

BTW thanks for the screenshot - I forgot to mention about it before.

**Comment by darkkatarsis** [comment]
![Uploading Screenshot 2020-10-07 at 00.15.45.png…]()
[log2.txt](https://github.com/maxwroc/battery-state-card/files/5337145/log2.txt)

**Comment by maxwroc** [comment]
One more try for today (v3)
[battery-state-card.zip](https://github.com/maxwroc/battery-state-card/files/5337321/battery-state-card.zip)

**Comment by darkkatarsis** [comment]
Looks like it worked. Can you tell me what the problem was?

If anything, I sent logs:
<img width="2556" alt="Screenshot 2020-10-07 at 08 26 58" src="https://user-images.githubusercontent.com/24244919/95295195-e967f700-0876-11eb-9169-7b329f77ecca.png">
ps. Even when I was hard refresh the site, I didn't notice "?"

log3.txt
[log3.txt](https://github.com/maxwroc/battery-state-card/files/5338608/log3.txt)

**Comment by maxwroc** [comment]
I'm glad to hear it worked! This issue was reported already by bunch of people (here and on HA community forum) and I was not able to fix it without having a repro. So thanks for your help!

TBH at the moment I don't know where exactly the problem is located but for sure it is caused by refreshing/re-rendering the card too often. And it looks like the final result is a mixture of two states which were passed for rendering. 

Surprisingly it affects only the icon - not other stuff. Maybe it is because SVG icons are slow to render and when the state is updated but previous icon haven't finished rendering the new update for the icon-item is discarded.

So to fix it I have added rendering throttling. It should improve the general performance as we are skipping now rendering previous states which would be overwritten anyway (millisecond later).

What next:
Today evening I will prepare the final version of the code. I would like to ask you to do one more test with the final code. Once I hear from you it works fine I will release new version of the card.

**Comment by darkkatarsis** [comment]
There is no problem. Just let me know in the commentary! :)

**Comment by maxwroc** [comment]
[battery-state-card.zip](https://github.com/maxwroc/battery-state-card/files/5342396/battery-state-card.zip)
Ok this is the final fix, please verify if the issue is gone and let me know

**Comment by darkkatarsis** [comment]
BATTERY-STATE-CARD  1.4.2 Working just fine!

**Comment by maxwroc** [comment]
Awesome. I think I will add yet another fix and I'll release 1.5.0 today/tomorrow with all of them
